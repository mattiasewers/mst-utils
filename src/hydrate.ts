import Store from './store';

export function dehydrate() {
  return JSON.stringify(Store.getJSON());
}

export function rehydrate() {
  return Store.create((<any>window).__STATE);
}
