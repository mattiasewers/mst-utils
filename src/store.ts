import { getSnapshot } from 'mobx-state-tree';

interface LooseObject {
  [key: string]: any;
}

class Store {
  private imports = {};
  public setup(stores) {
    if (!stores) {
      throw new Error('Must provide atleast 1 mst model.');
    }
    this.imports = stores;
    return this;
  }

  public create(state = {}) {
    if (Object.keys(this.imports).length === 0) {
      throw new Error('You must run the setup function before create.');
    }
    Object.keys(this.imports).forEach(key => {
      const StoreModel = this.imports[key];
      try {
        const instance = StoreModel.create(state[key]);
        this[key] = instance;
      } catch (err) {
        throw new Error(err);
      }
    });
    return this;
  }

  public getJSON() {
    const stores = {};
    Object.keys(this.imports).forEach((key: string) => {
      const instance = this[key];
      stores[key] = getSnapshot(instance);
    });
    return stores;
  }

  public destroy() {
    Object.keys(this.imports).forEach((key: string) => {
      delete this.imports[key];
      delete this[key];
    });
  }
}

export default new Store() as Store & LooseObject;
