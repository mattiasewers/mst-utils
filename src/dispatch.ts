import store from './store';
import get from 'lodash.get';

export default function dispatch(opt, ...args) {
  const fn = get(store, opt);
  if (typeof fn !== 'function') {
    throw new Error('Not a function');
  }

  return fn(...args);
}
