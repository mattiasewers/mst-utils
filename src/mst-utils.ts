import store from './store';
import dispatch from './dispatch';
import { rehydrate, dehydrate } from './hydrate';

export { store, dispatch, rehydrate, dehydrate };
