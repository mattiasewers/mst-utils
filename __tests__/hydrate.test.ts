import { types } from 'mobx-state-tree';
import Store from '../src/store';
import { dehydrate, rehydrate } from '../src/hydrate';

const count = types
  .model('Count', {
    text: types.string,
    tags: types.optional(types.array(types.string), ['foo'])
  })
  .actions(self => ({
    change(val) {
      self.text = val;
    },
    changeType(val) {
      self.tags = val;
    }
  }));

describe('hydrate.ts', () => {
  it('should dehydrate', () => {
    const store = Store.setup({ count });
    const state = { count: { text: 'hello' } };
    store.create(state);
    expect(dehydrate()).toMatchSnapshot();
    store.destroy();
  });

  it('should rehydrate', () => {
    const store = Store.setup({ count });
    const state = { count: { text: 'hello' } };
    (window as any).__STATE = state;

    rehydrate();

    expect(store.count.text).toBe('hello');
    store.destroy();
  });
});
