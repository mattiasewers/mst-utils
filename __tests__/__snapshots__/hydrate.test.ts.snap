// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hydrate.ts should dehydrate 1`] = `"{\\"count\\":{\\"text\\":\\"hello\\",\\"tags\\":[\\"foo\\"]}}"`;
